cms/search/publishers.py

```diff
@@ -0,0 +1,202 @@
+import json
+import logging
+from abc import ABC, abstractmethod
+from typing import TYPE_CHECKING, Optional
+
+from django.conf import settings
+from kafka import KafkaProducer
+
+logger = logging.getLogger(__name__)
+
+
+if TYPE_CHECKING:
+    from kafka.producer.future import RecordMetadata
+    from wagtail.models import Page
+
+
+class BasePublisher(ABC):
+    """BasePublisher defines shared functionalities, such as how to build a message
+    for created/updated or deleted events. Each subclass only needs to define how
+    to actually publish (i.e., send) the built message.
+
+    dp-search-data-extractor spec link:
+    https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/specification.yml#L53
+    dp-search-data-importer spec link:
+    https://github.com/ONSdigital/dp-search-data-importer/blob/30fb507e90f2cf1974ec0ca43bb0466307e2f112/specification.yml#L186
+    contract: https://github.com/ONSdigital/dis-search-upstream-stub/blob/main/docs/contract/resource_metadata.yml
+    """
+
+    def publish_created_or_updated(self, page: "Page") -> None:
+        """Build the message for the created/updated event.
+        Delegate sending to the subclass's _publish_to_service().
+        """
+        topic = self.get_topic_created_or_updated()
```

Comment:
nitpick: while this follows the terminology used in the dp-search schema, I think we should adapt it to our project. Speficially, we have `Topic` from taxonomy, and Topic page. But here "topic" is the operation https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/specification.yml#L13-L50

Let's make this relevant, and avoid confusion. So `operation` and `get_created_or_updated_operation`. Or worst case scenarion `channel` (https://github.com/ONSdigital/dp-search-data-extractor/blob/develop/specification.yml#L13-L28)

Reply:
Let's go with channel

Reply:
Done!